src/main/java/org/opensearch/plugin/insights/core/service/categorizer/QueryShapeGenerator.java

```diff
@@ -191,15 +311,113 @@ static String buildSortShape(List<SortBuilder<?>> sortBuilderList, Boolean showF
     }
 
     /**
-     * Method to build field data
-     * @return String: comma separated list with leading space in square brackets
+     * Builds a field data string from a builder.
+     *
+     * @param builder                  aggregation or sort builder
+     * @param propertiesAsMap       properties
+     * @param index                 index
+     * @param showFieldName            whether to include field names
+     * @param showFieldType            whether to include field types
+     * @return Field data string
      * Ex: " [my_field, width:5]"
      */
-    static String buildFieldDataString(NamedWriteable builder) {
+    String buildFieldDataString(
+        NamedWriteable builder,
+        Map<String, Object> propertiesAsMap,
+        Index index,
+        Boolean showFieldName,
+        Boolean showFieldType
+    ) {
         List<String> fieldDataList = new ArrayList<>();
         if (builder instanceof WithFieldName) {
-            fieldDataList.add(((WithFieldName) builder).fieldName());
+            String fieldName = ((WithFieldName) builder).fieldName();
+            if (showFieldName) {
+                fieldDataList.add(fieldName);
+            }
+            if (showFieldType) {
+                String fieldType = getFieldType(fieldName, propertiesAsMap, index);
+                if (fieldType != null && !fieldType.isEmpty()) {
+                    fieldDataList.add(fieldType);
+                }
+            }
         }
         return " [" + String.join(", ", fieldDataList) + "]";
     }
+
+    String getFieldType(String fieldName, Map<String, Object> propertiesAsMap, Index index) {
+        String fieldType = getFieldTypeFromCache(fieldName, index);
+        if (fieldType != null) {
+            if (fieldType.equals(NO_FIELD_TYPE_VALUE)) {
+                return null;
+            } else {
+                return fieldType;
+            }
+        }
+
+        fieldType = getFieldTypeFromMapping(index, fieldName, propertiesAsMap);
+        if (fieldType != null) {
+            String finalFieldType = fieldType;
+            fieldTypeMap.computeIfAbsent(index, k -> new ConcurrentHashMap<>()).computeIfAbsent(fieldName, k -> finalFieldType);
+            return fieldType;
+        }
+        return null;
+    }
+
+    String getFieldTypeFromMapping(Index index, String fieldName, Map<String, Object> propertiesAsMap) {
+        // Recursively find the field type from properties
+        if (propertiesAsMap != null) {
+            String fieldType = findFieldTypeInProperties(propertiesAsMap, fieldName.split("\\."), 0);
+
+            // Cache the NO_FIELD_TYPE_VALUE result if no field type is found in this index
+            if (fieldType == null) {
+                fieldTypeMap.computeIfAbsent(index, k -> new ConcurrentHashMap<>()).computeIfAbsent(fieldName, k -> fieldType);
+            }
```

Comment:
The remapping function evaluates to null here, hence nothing will be added to the map